agent/local/state.go

```diff
@@ -0,0 +1,972 @@
+package local
+
+import (
+	"fmt"
+	"log"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/agent/token"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/lib"
+	"github.com/hashicorp/consul/types"
+)
+
+// permissionDenied is returned when an ACL based rejection happens.
+const permissionDenied = "Permission denied"
+
+// Config is the configuration for the State. It is
+// populated during NewLocalAgent from the agent configuration to avoid
+// race conditions with the agent configuration.
+type Config struct {
+	AdvertiseAddr       string
+	CheckUpdateInterval time.Duration
+	Datacenter          string
+	NodeID              types.NodeID
+	NodeName            string
+	TaggedAddresses     map[string]string
+}
+
+// ServiceState describes the state of a service record.
+type ServiceState struct {
+	// Service is the local copy of the service record.
+	Service *structs.NodeService
+
+	// Token is the ACL to update the service record on the server.
+	Token string
+
+	// InSync contains whether the local state of the service record
+	// is in sync with the remote state on the server.
+	InSync bool
+
+	// Deleted is true when the service record has been marked as deleted
+	// but has not been removed on the server yet.
+	Deleted bool
+}
+
+// Clone returns a shallow copy of the object. The service record still
+// points to the original service record and must not be modified.
+func (s *ServiceState) Clone() *ServiceState {
+	s2 := new(ServiceState)
+	*s2 = *s
+	return s2
+}
+
+// CheckState describes the state of a health check record.
+type CheckState struct {
+	// Check is the local copy of the health check record.
+	Check *structs.HealthCheck
+
+	// Token is the ACL record to update the health check record
+	// on the server.
+	Token string
+
+	// CriticalTime is the last time the health check status went
+	// from non-critical to critical. When the health check is not
+	// in critical state the value is the zero value.
+	CriticalTime time.Time
+
+	// DeferCheck is used to delay the sync of a health check when
+	// only the status has changed.
```

Comment:
It's actually for when the status _has not_ changed but there has been a change to the output. It's used to rate limit changes like that which don't affect the state of the node/service since the status didn't update.

Reply:
done